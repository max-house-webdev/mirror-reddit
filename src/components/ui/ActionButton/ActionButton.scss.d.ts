export const actionButton: string;
export const actionButtonText: string;
export const rotate180: string;
export const translate: string;
