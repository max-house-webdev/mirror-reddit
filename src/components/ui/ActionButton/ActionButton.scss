@import '../../../assets/scss/base/colors.scss';

.actionButton {
  &:focus-visible {
    outline: $orange 1px solid;
    outline-offset: 2px;
  }

  &:hover {
    background-color: $greyEC;

    &Text {
      color: $orange;
    }
  }

  & > svg {
    width: 20px;
    height: 20px;
  }

  &Text {
    display: inline-flex;
    align-items: center;
    color: $greyC4;
    margin-right: 1.75rem;

    & > svg {
      margin-right: 0.375rem;
    }
  }
}

.rotate180 {
  & > svg {
    transform: rotate(180deg) translateY(4px);
  }
}

.translate {
  & > svg {
    transform: translateY(6px);
  }
}
