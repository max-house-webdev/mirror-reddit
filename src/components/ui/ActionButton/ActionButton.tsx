import React from 'react';
import classNames from 'classnames';

import styles from './ActionButton.scss';

import {
  IconSave,
  IconShare,
  IconComplain,
  IconComment,
  IconDelete,
  IconTriangleUp,
} from '../../../assets/svg';

type TAction =
  | 'share'
  | 'save'
  | 'answer'
  | 'complain'
  | 'delete'
  | 'count'
  | 'update post list'
  | 'load all posts';

export interface IButtonProps {
  type?: 'button' | 'submit' | 'reset';
  action: TAction;
  textContent?: boolean;
  rotate?: boolean;
  translate?: boolean;
  onClick: React.MouseEventHandler<HTMLButtonElement>;
}

export function ActionButton(props: IButtonProps) {
  const {
    type = 'button',
    action,
    onClick,
    textContent = true,
    rotate = false,
    translate = false,
  } = props;

  let role: Record<string, JSX.Element | string | null> = {};

  switch (action) {
    case 'share':
      role = { icon: <IconShare />, text: 'Поделиться' };
      break;
    case 'save':
      role = { icon: <IconSave />, text: 'Сохранить' };
      break;
    case 'answer':
      role = { icon: <IconComment />, text: 'Ответить' };
      break;
    case 'complain':
      role = { icon: <IconComplain />, text: 'Пожаловаться' };
      break;
    case 'delete':
      role = { icon: <IconDelete />, text: 'Удалить' };
      break;
    case 'count':
      role = { icon: <IconTriangleUp />, text: '' };
      break;
  }

  const classes = classNames(
    styles['actionButton'],
    {
      [styles['actionButtonText']]: textContent,
    },
    {
      [styles['rotate180']]: rotate,
    },
    {
      [styles['translate']]: translate,
    }
  );

  const { icon } = role;
  const text = textContent ? role.text : '';

  return (
    <button type={type} className={classes} onClick={onClick}>
      {icon}
      {text}
    </button>
  );
}
