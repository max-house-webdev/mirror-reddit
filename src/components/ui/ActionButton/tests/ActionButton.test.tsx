import React from 'react';
import { shallow } from 'enzyme';

import { ActionButton } from '../ActionButton';

// empty handleClick
// eslint-disable-next-line @typescript-eslint/no-empty-function
const handleClick = () => {};

describe('<ActionButton />', () => {
  describe('should render with textContent', () => {
    test('type=button class=Button Button_text textContent=Поделиться', () => {
      const wrapper = shallow(
        <ActionButton action={'share'} onClick={handleClick} />
      );

      expect(wrapper.find('button.Button.Button_text')).toBeDefined();
      expect(wrapper.find('button').text()).toContain('Поделиться');

      expect(wrapper).toMatchSnapshot();
    });

    test('type=button class=Button Button_text textContent=Сохранить', () => {
      const wrapper = shallow(
        <ActionButton action={'save'} onClick={handleClick} />
      );

      expect(wrapper.find('button').text()).toContain('Сохранить');

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('should render without textContent', () => {
    test('type=button class=Button', () => {
      const wrapper = shallow(
        <ActionButton
          action={'delete'}
          textContent={false}
          onClick={handleClick}
        />
      );

      expect(wrapper.find('button').text().length).toBe(0);

      expect(wrapper).toMatchSnapshot();
    });
  });
});
