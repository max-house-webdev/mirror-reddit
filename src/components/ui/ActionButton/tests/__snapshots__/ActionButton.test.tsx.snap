// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionButton /> should render with textContent type=button class=Button Button_text textContent=Поделиться 1`] = `
<button
  className="undefined"
  onClick={[Function]}
  type="button"
>
  <test-file-stub />
  Поделиться
</button>
`;

exports[`<ActionButton /> should render with textContent type=button class=Button Button_text textContent=Сохранить 1`] = `
<button
  className="undefined"
  onClick={[Function]}
  type="button"
>
  <test-file-stub />
  Сохранить
</button>
`;

exports[`<ActionButton /> should render without textContent type=button class=Button 1`] = `
<button
  className=""
  onClick={[Function]}
  type="button"
>
  <test-file-stub />
</button>
`;
